Sources/MongoSwift/ChangeStream.swift

```diff
@@ -84,38 +84,11 @@ public class ChangeStream<T: Codable>: Sequence, IteratorProtocol {
     /// `maxAwaitTimeMS` milliseconds as specified in the `ChangeStreamOptions`, or for the server default timeout
     /// if omitted.
     public func next() -> T? {
-        // If an error exists, refuse iterating the change stream to avoid overwriting the original error.
-        guard self.error == nil else {
-            return nil
-        }
-        // Allocate space for a reference to a BSON pointer.
-        let out = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
-        defer {
-            out.deinitialize(count: 1)
-            out.deallocate()
-        }
-
-        guard mongoc_change_stream_next(self.changeStream, out) else {
-            return nil
-        }
-
-        guard let pointee = out.pointee else {
-            fatalError("The change stream was advanced, but the document is nil.")
-        }
-
-        // we have to copy because libmongoc owns the pointer.
-        let doc = Document(copying: pointee)
-
-        // Update the resumeToken with the `_id` field from the document.
-        guard let resumeToken = doc["_id"] as? Document else {
-            self.swiftError =
-                    RuntimeError.internalError(message: "_id field is missing from the change stream document.")
-            return nil
-        }
-        self.resumeToken = ResumeToken(resumeToken)
-
         do {
-            return try decoder.decode(T.self, from: doc)
+            let operation = ChangeStreamNextOperation(changeStream: self)
+            let out = try operation.execute(using: self.connection, session: self.session)
+            self.swiftError = nil
```

Comment:
Yeah, I can't think of a situation in which we would want to clear an error that we've encountered. Also, I think the `guard` that was moved into the operation should just remain in `next` at the change stream level, so that no operation will even be created in the event that an error has been seen. Agreed on automatically closing.

Reply:
I closed the `ChangeStream` if an error was encountered, but should I also close in the case that there are no documents returned and an error occurred / the cursor can't possibly return any more data (unsure of the situations where this happens)? I am assuming that a `ChangeStream` operates like a tailable cursor (which might be wrong) so am looking at cases 2 and 3 for when the `next` operation of `MongoCursor` doesn't return a doc.

Reply:
yup, you are right that change streams behave like tailable cursors.

so case 2 where we encountered some error iterating a change stream is definitely applicable here.
case 3 for tailable cursors is something that we can encounter in cases like [this](https://github.com/mongodb/mongo-swift-driver/blob/master/Tests/MongoSwiftTests/MongoCursorTests.swift#L61). if nothing matches the initial tailable cursor query (in the test that occurs because the collection is empty) then the server will not keep the cursor alive and `mongoc_cursor_more` would return false.

I don't think case 3 is relevant for change streams as the server won't close a change stream even if it is initially empty. and mongoc provides us no way to check something like `mongoc_cursor_more` anyway.

if we're going to make this closing change now, I think we should switch `ChangeStream` to store all of its properties in a state enum much like `MongoCursor` does, so that we can nil everything out at once when we close it. and we probably need to/should save off errors similarly to how we do there rather than recomputing each time.
right now we have a couple problems in that we will potentially call `close()` (and therefore `mongoc_change_stream_destroy`) twice, once via manual `close` and once via `deinit`, and in that after we call `close` we can't access any mongoc error that was encountered after calling `close()`.